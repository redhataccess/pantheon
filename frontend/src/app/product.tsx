import React, { Component } from 'react';
import { Bullseye, Button, Alert, AlertActionCloseButton, Form, FormGroup, TextInput, ActionGroup } from '@patternfly/react-core';
import '@app/app.css';
import { Redirect } from 'react-router-dom'

class Product extends Component {
  public state = {
    failedPost: false,
    isDup: false,
    isMissingFields: false,
    login: false,
    productDescription: '',
    productName: '',
    redirect: false,
    results: [],
  };

  // render method transforms the react components into DOM nodes for the browser.
  public render() {
    const { productName, productDescription, isMissingFields, isDup } = this.state;
    return (
      <React.Fragment>
        {/* Bullseye makes sure everyhting is in the middle */}
        <Bullseye>
          <Form>
            <div className="app-container">
              <div>
                {isMissingFields && (
                  <div className="notification-container">
                    <Alert
                      variant="warning"
                      title="A Product name is required."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    />
                  </div>
                )}
                {isDup && (
                  <div className="notification-container">
                    <Alert
                      variant="warning"
                      title="Duplicated Product name."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    />
                  </div>
                )}
                <FormGroup
                  label="Product Name"
                  isRequired={true}
                  fieldId="product-name"
                >
                  <TextInput isRequired={true} id="product-name" type="text" placeholder="Product Name" value={productName} onChange={this.handleNameInput} />
                </FormGroup>
                <br />
                <FormGroup
                  label="Product Description"
                  fieldId="product-description"
                >
                  <TextInput id="product-description" type="text" placeholder="Product Description" value={productDescription} onChange={this.handleProductInput} />
                </FormGroup>
                <br />
                <ActionGroup>
                  <Button aria-label="Creates a new Product Name with Description specified." onClick={this.saveProduct}>Save</Button>
                  <div>
                    {this.checkAuth()}
                    {this.loginRedirect()}
                    {this.renderRedirect()}
                  </div>
                </ActionGroup>
              </div>
            </div>
          </Form>
        </Bullseye>
      </React.Fragment>
    );
  }
  // methods that handle the state changes.
  private handleNameInput = productName => {
    this.setState({ productName });
  };

  private handleProductInput = productDescription => {
    this.setState({ productDescription });
    // console.log("Desc " + productDescription)
  };

  private saveProduct = () => {
    if (this.state.productName === '') {
      this.setState({ isMissingFields: true })
    } else {
      this.productExist().then(exist => {
        if (!exist) {
          const hdrs = {
            'Accept': 'application/json',
            'cache-control': 'no-cache'
          }
          // setup url fragment
          const urlFragment = this.state.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
          const formData = new FormData();
          formData.append("name", this.state.productName)
          formData.append("description", this.state.productDescription)
          formData.append("sling:resourceType", "pantheon/product")
          formData.append("jcr:primaryType", 'pant:product')
          // currently we don't translate products in Customer Portal.
          formData.append("locale", "en-US")
          formData.append("url", urlFragment)
          // fetch makes the request to create a new product.
          // transfor productName to lower case and replace special chars with _.
          fetch(encodeURI('/content/products/' + urlFragment), {
            body: formData,
            headers: hdrs,
            method: 'post'
          }).then(response => {
            if (response.status === 201 || response.status === 200) {
              this.setState({ redirect: true })
            } else if (response.status === 500) {
              // console.log(" Needs login " + response.status)
              this.setState({ login: true })
            } else {
              console.log(" Failed " + response.status)
              this.setState({ failedPost: true })
            }
          });
        }
      }
      )
    }
  }

  private renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to='/products' />
    } else {
      return ""
    }
  }

  private loginRedirect = () => {
    if (this.state.login) {
      return <Redirect to='/login' />
    } else {
      return ""
    }
  }

  private checkAuth = () => {
    fetch("/system/sling/info.sessionInfo.json")
      .then(response => response.json())
      .then(responseJSON => {
        const key = "userID"
        if (responseJSON[key] === 'anonymous') {
          this.setState({ login: true })
        }
      })
  }

  private dismissNotification = () => {
    this.setState({ isMissingFields: false, isDup: false });
  }

  private productExist = () => {
    let exists = false
    const backend = '/content/products/' + this.state.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_") + '.json'
    return fetch(backend)
      .then(response => {
        if (response.status === 200) {
          exists = true
          this.setState({ isDup: true })
        }
      }).then(() => exists)
  }
}

export { Product }