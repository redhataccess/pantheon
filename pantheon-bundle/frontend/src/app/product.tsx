import React, { Component } from 'react'
import { Bullseye, Button, Alert, AlertActionCloseButton, Form, FormGroup, TextInput, ActionGroup } from '@patternfly/react-core'
import '@app/app.css'
import { Redirect } from 'react-router-dom'

interface IState {
  failedPost: boolean
  isDup: boolean
  isMissingFields: boolean
  productDescription: string
  productName: string
  redirect: boolean
}

class Product extends Component<any, IState> {
  constructor(props) {
    super(props)
    this.state = {
      failedPost: false,
      isDup: false,
      isMissingFields: false,
      productDescription: '',
      productName: '',
      redirect: false
    }

  }

  // render method transforms the react components into DOM nodes for the browser.
  public render() {
    return (
      <React.Fragment>
        {/* Bullseye makes sure everyhting is in the middle */}
        <Bullseye>
          <Form>
            <div className="app-container">
              <div>
                {this.state.isMissingFields && 
                  <div className="notification-container">
                    <Alert
                      variant="warning"
                      title="A Product name is required."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    />
                  </div>
                }
                {this.state.isDup && 
                  <div className="notification-container">
                    <Alert
                      variant="warning"
                      title="Duplicated Product name."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    />
                  </div>
                }
                {this.state.failedPost &&
                  <div className="notification-container">
                    <Alert
                      variant="danger"
                      title="Failed to create product."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    >
                    Please check if you are logged in as a publisher.
                    </Alert>
                  </div>
                }
                <FormGroup
                  label="Product Name"
                  isRequired={true}
                  fieldId="product-name"
                >
                  <TextInput isRequired={true} id="product-name" type="text" placeholder="Product Name" value={this.state.productName} onChange={this.handleNameInput} />
                </FormGroup>
                <br />
                <FormGroup
                  label="Product Description"
                  fieldId="product-description"
                >
                  <TextInput id="product-description" type="text" placeholder="Product Description" value={this.state.productDescription} onChange={this.handleProductInput} />
                </FormGroup>
                <br />
                <ActionGroup>
                  <Button aria-label="Creates a new Product Name with Description specified." onClick={this.saveProduct}>Save</Button>
                  <div>
                    {this.renderRedirect()}
                  </div>
                </ActionGroup>
              </div>
            </div>
          </Form>
        </Bullseye>
      </React.Fragment>
    )
  }
  // methods that handle the state changes.
  private handleNameInput = productName => {
    this.setState({ productName })
  }

  private handleProductInput = productDescription => {
    this.setState({ productDescription })
    // console.log("Desc " + productDescription)
  }

  private saveProduct = () => {
    if (this.state.productName === '') {
      this.setState({ isMissingFields: true })
    } else {
      this.productExist().then(exist => {
        if (!exist) {
          const hdrs = {
            'Accept': 'application/json',
            'cache-control': 'no-cache'
          }
          // setup url fragment
          const urlFragment = this.state.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_")
          const formData = new FormData()
          formData.append("name", this.state.productName)
          formData.append("description", this.state.productDescription)
          formData.append("sling:resourceType", "pantheon/product")
          formData.append("jcr:primaryType", 'pant:product')
          // currently we don't translate products in Customer Portal.
          formData.append("locale", "en-US")
          formData.append("url", urlFragment)
          // fetch makes the request to create a new product.
          // transfor productName to lower case and replace special chars with _.
          fetch(encodeURI('/content/products/' + urlFragment), {
            body: formData,
            headers: hdrs,
            method: 'post'
          }).then(response => {
            if (response.status === 201 || response.status === 200) {
              this.setState({ redirect: true })
            } else {
              this.setState({ failedPost: true })
            }
          })
        }
      }
      )
    }
  }

  private renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to='/products' />
    } else {
      return ""
    }
  }

  private dismissNotification = () => {
    this.setState({ isMissingFields: false, isDup: false })
  }

  private productExist = () => {
    let exists = false
    const backend = '/content/products/' + this.state.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_") + '.json'
    return fetch(backend)
      .then(response => {
        if (response.status === 200) {
          exists = true
          this.setState({ isDup: true })
        }
      }).then(() => exists)
  }
}

export { Product }