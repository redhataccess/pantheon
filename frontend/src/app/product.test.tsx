import React from 'react';
import { Product } from '@app/product';
import "isomorphic-fetch"

import { mount, shallow } from 'enzyme';
import { Bullseye, TextInput, FormGroup, Button } from '@patternfly/react-core';
import renderer from 'react-test-renderer';

describe('Product tests', () => {
  test('should render Product component', () => {
    const view = shallow(<Product />);
    expect(view).toMatchSnapshot();
  });

  it('should render a Bullseye layout', () => {
    const wrapper = mount(<Product />);
    const bullseyeLayout = wrapper.find(Bullseye);
    expect(bullseyeLayout.exists()).toBe(true)
  });


  it('should render a form group', () => {
    const wrapper = mount(<Product />);
    const formGroup = wrapper.find(FormGroup);
    expect(formGroup.exists()).toBe(true)
  });

  it('should render a text input', () => {
    const wrapper = mount(<Product />);
    const textInput = wrapper.find(TextInput);
    expect(textInput.exists()).toBe(true)
  });

  it('should render a Button', () => {
    const wrapper = mount(<Product />);
    const button = wrapper.find(Button);
    expect(button.exists()).toBe(true)
  });

  it('test renderRedirect function', () => {
    const wrapper = renderer.create(<Product />);
    const inst = wrapper.getInstance();
    expect(inst.renderRedirect).toMatchSnapshot();
  });

  it('test loginRedirect function', () => {
    const wrapper = renderer.create(<Product />);
    const inst = wrapper.getInstance();
    expect(inst.loginRedirect).toMatchSnapshot();
  });
 
  it('test checkAuth function', () => {
    const wrapper = renderer.create(<Product />);
    const inst = wrapper.getInstance();
    expect(inst.checkAuth).toMatchSnapshot();
  });
  
  it('test dismissNotification function', () => {
    const wrapper = renderer.create(<Product />);
    const inst = wrapper.getInstance();
    expect(inst.dismissNotification).toMatchSnapshot();
  });

  it('test productExist function', () => {
    const wrapper = renderer.create(<Product />);
    const inst = wrapper.getInstance();
    expect(inst.productExist("Red Hat Enterprise Linux")).toMatchSnapshot();
  });

  it('test handleNameInput function', () => {
    const wrapper = renderer.create(<Product />);
    const inst = wrapper.getInstance();
    expect(inst.handleNameInput("Red Hat Enterprise Linux")).toMatchSnapshot();
  });

  it('test handleProductInput function', () => {
    const wrapper = renderer.create(<Product />);
    const inst = wrapper.getInstance();
    expect(inst.handleProductInput("Linux Platform")).toMatchSnapshot();
  });
});
