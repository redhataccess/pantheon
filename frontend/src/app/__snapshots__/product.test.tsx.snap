// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product tests should render Product component 1`] = `
<Fragment>
  <Bullseye>
    <Form>
      <div
        className="app-container"
      >
        <div>
          <FormGroup
            fieldId="product-name"
            isRequired={true}
            label="Product Name"
          >
            <TextInput
              aria-label={null}
              className=""
              id="product-name"
              isDisabled={false}
              isReadOnly={false}
              isRequired={true}
              isValid={true}
              onChange={[Function]}
              placeholder="Product Name"
              type="text"
              value=""
            />
          </FormGroup>
          <br />
          <FormGroup
            fieldId="product-description"
            label="Product Description"
          >
            <TextInput
              aria-label={null}
              className=""
              id="product-description"
              isDisabled={false}
              isReadOnly={false}
              isRequired={false}
              isValid={true}
              onChange={[Function]}
              placeholder="Product Description"
              type="text"
              value=""
            />
          </FormGroup>
          <br />
          <ActionGroup>
            <Button
              aria-label="Creates a new Product Name with Description specified."
              onClick={[Function]}
            >
              Save
            </Button>
            <div />
          </ActionGroup>
        </div>
      </div>
    </Form>
  </Bullseye>
</Fragment>
`;

exports[`Product tests test checkAuth function 1`] = `[Function]`;

exports[`Product tests test dismissNotification function 1`] = `[Function]`;

exports[`Product tests test handleNameInput function 1`] = `undefined`;

exports[`Product tests test handleProductInput function 1`] = `undefined`;

exports[`Product tests test loginRedirect function 1`] = `[Function]`;

exports[`Product tests test productExist function 1`] = `Promise {}`;

exports[`Product tests test renderRedirect function 1`] = `[Function]`;
